modules/twinklespeedy.js

```diff
@@ -1010,6 +1013,156 @@ Twinkle.speedy.initDialog = function twinklespeedyInitDialog(callbackfunc) {
+	// This is not optimal at all, and if we could scan the CSD code instead of making it an ugly custom thing
+	// I'd be all for it
+	const values = ['reason'];
+
+	const templateParams = [[]];
+	templateParams[0][1] = csdReason;
+
+	const normalizeds = values.map((value) => Twinkle.speedy.normalizeHash[value]);
+
+	// analyse each criterion to determine whether to watch the page, prompt for summary, or notify the creator
+	let watchPage, promptForSummary;
+	normalizeds.forEach((norm) => {
+		if (Twinkle.getPref('watchSpeedyPages').includes(norm)) {
+			watchPage = Twinkle.getPref('watchSpeedyExpiry');
+		}
+		if (Twinkle.getPref('promptForSpeedyDeletionSummary').includes(norm)) {
+			promptForSummary = true;
+		}
+	});
+
+	const warnusertalk = !Twinkle.speedy.hasCSD && normalizeds.some((norm, index) => Twinkle.getPref('warnUserOnSpeedyDelete').includes(norm) &&
+			!(norm === 'g6' && values[index] !== 'copypaste') && !(norm === 'g5' && values[index] !== 'gs'));
+
+	const welcomeuser = warnusertalk && normalizeds.some((norm) => Twinkle.getPref('welcomeUserOnSpeedyDeletionNotification').includes(norm));
+
+	const params = {
+		values: values,
+		normalizeds: normalizeds,
+		watch: watchPage,
+		deleteTalkPage: Twinkle.getPref('deleteTalkPageOnDelete'),
+		deleteRedirects: Twinkle.getPref('deleteRedirectsOnDelete'),
+		warnUser: warnusertalk,
+		welcomeuser: welcomeuser,
+		promptForSummary: promptForSummary,
+		templateParams: templateParams
+	};
+
+	Twinkle.speedy.callbacks.sysop.main(params);
+};
+
+Twinkle.speedy.instantDecline = function twinklespeedyInstantDecline() {
+	const wikipediaPage = new Morebits.wiki.Page(mw.config.get('wgPageName'), 'Tagging page');
+	wikipediaPage.load(Twinkle.speedy.instantDecline.main);
+};
+
+Twinkle.speedy.instantDecline.main = function twinklespeedyInstantDeclineMain(pageObj) {
+	// finds the db template
+	const text = pageObj.getPageText();
+	const tag = /(?:\{\{\s*(db|delete|db-.*?|speedy deletion-.*?)(?:\s*\||\s*\}\}))/.exec(text)[1];
+	let wikitextObj = new Morebits.wikitext.page(text);
+
+	// and then removes it
+	wikitextObj.removeTemplate(tag);
+
+	if (wikitextObj.getText()[0] === '\n') {
+		wikitextObj = new Morebits.wikitext.page(wikitextObj.getText().slice(1));
+	}
+
+	pageObj.setPageText(wikitextObj.getText());
+
+	const splitTag = tag.split(/-(.*)/).length > 1 ? tag.split(/-(.*)/)[1] : tag;
+
+	if (splitTag in Twinkle.speedy.normalizeHash) {
+		if (Twinkle.speedy.normalizeHash[splitTag] === 'db') {
+				pageObj.setEditSummary('Declining speedy deletion (no code provided).');
+		} else {
+				pageObj.setEditSummary('Declining speedy deletion ([[WP:CSD#' + Twinkle.speedy.normalizeHash[splitTag].toUpperCase() + '|CSD ' + Twinkle.speedy.normalizeHash[splitTag].toUpperCase() + ']]).');
+		}
+	} else {
+		pageObj.setEditSummary('Declining speedy deletion (multiple criteria).');
+	}
+	pageObj.setChangeTags(Twinkle.changeTags);
```

Comment:
This doesn't handle watchlisting, which should be based on `watchSpeedyPages` pref, if not a new pref.